src/webapi/web_socket.rs

```diff
@@ -0,0 +1,287 @@
+        BinaryType::from_str(&binary_type)
+    }
+
+    /// Sets the binary type of the web socket. Only affects received messages.
+    /// The default binary type is `Blob`.
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket)
+    pub fn set_binary_type(&self, binary_type: BinaryType) {
+        js!( @(no_return) @{self}.binaryType = @{binary_type.to_str()}; );
+    }
+
+    /// Returns the number of bytes of data that have been queued using calls to send()
+    /// but not yet transmitted to the network. This value resets to zero once all queued
+    /// data has been sent. This value does not reset to zero when the connection is closed;
+    /// if you keep calling send(), this will continue to climb. 
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket)
+    pub fn buffered_amount(&self) -> u64 {
+        js!( return @{self}.bufferedAmount; ).try_into().unwrap()
+    }
+
+    /// Returns the extensions selected by the server. This is currently only the empty
+    /// string or a list of extensions as negotiated by the connection.
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket)
+    pub fn extensions(&self) -> String {
+        js!( return @{self}.extensions; ).try_into().unwrap()
+    }
+
+    /// Returns a string indicating the name of the sub-protocol the server selected;
+    /// this will be one of the strings specified in the protocols parameter when
+    /// creating the WebSocket object.
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket)
+    pub fn protocol(&self) -> String {
+        js!( return @{self}.protocol; ).try_into().unwrap()
+    }
+
+    /// Returns the current state of the connection.
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket)
+    pub fn ready_state(&self) -> WebSocketReadyState {
+        js!( return @{self}.protocol; ).try_into().unwrap()
+    }
+
+    /// Returns the URL as resolved by the constructor. This is always an absolute URL.
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket)
+    pub fn url(&self) -> String {
+        js!( return @{self}.url; ).try_into().unwrap()
+    }
+
+    /// Closes the WebSocket connection or connection attempt, if any. If the connection
+    /// is already CLOSED, this method does nothing.
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket#close())
+    pub fn close(&self) {
+        js!( @(no_return) @{self}.close(); );
+    }
+
+    /// Closes the WebSocket connection or connection attempt, if any. If the connection
+    /// is already CLOSED, this method does nothing.
+    ///
+    /// [(JavaScript docs)](https://developer.mozilla.org/en-US/docs/Web/API/WebSocket#close())
+    pub fn close_with_status(&self, code: WebSocketCloseCode, reason: &str) {
```

Comment:
@Herschel is implementing proper exception support, I will add a temporary error type here for the moment rather than try to map it intelligently.

Reply:
I'm fine with not handling it properly right now, as long as it returns a `Result` with a private error type so that we won't have to change the prototype later.